File: p2/p208/Trie.java
package leetcode.p2.p208;


public class Trie {

    private TrieNode root;

    /**
     * Time complexity: O(1)
     * Space complexity: O(1)
     */
    public Trie() {
        root = new TrieNode();
    }

    /**
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    public void insert(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (cur.children[index] == null) {
                cur.children[index] = new TrieNode();
            }
            cur = cur.children[index];
        }
        cur.isWord = true;
    }

    /**
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public boolean search(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (cur.children[index] == null) {
                return false;
            }
            cur = cur.children[index];
        }
        return cur.isWord;
    }

    /**
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for (int i = 0; i < prefix.length(); i++) {
            int index = prefix.charAt(i) - 'a';
            if (cur.children[index] == null) {
                return false;
            }
            cur = cur.children[index];
        }
        return true;
    }
}

class TrieNode {
    boolean isWord;
    TrieNode[] children;

    public TrieNode() {
        isWord = false;
        children = new TrieNode[26];
    }
}

File: p2/p208/trie.py
class Trie:
    """
    Feb 23, 2025 00:29
    """
    class TrieNode:
        def __init__(self):
            self.is_word = False
            self.children = {}

    def __init__(self):
        self.root = self.TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = self.TrieNode()
            curr = curr.children[c]

        curr.is_word = True

    def search(self, word: str) -> bool:
        curr = self.root
        for c in word:
            if c in curr.children:
                curr = curr.children[c]
            else:
                return False

        return curr.is_word

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for c in prefix:
            if c in curr.children:
                curr = curr.children[c]
            else:
                return False

        return True


if __name__ == '__main__':
    trie = Trie()

    trie.insert('apple')
    print(trie.search('apple') == True)
    print(trie.search('app') == False)
